---
title: "iBET - interactive Bioconductor Exploratory Tools"
author: 
- name: "Jared Andrews"
  email: jared.andrews@stjude.org
  affiliation: St. Jude Children's Research Hospital, Memphis, TN
date: "September 20th, 2021"
output: 
  html_document:
    code_folding: show
    toc: true
    toc_depth: 5
    toc_float: true	 
runtime: shiny
vignette: >
  %\VignetteIndexEntry{iBET - Interactive Shiny Widgets for Bioconductor Packages}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{css, echo=FALSE}
# This is to change width of the content so that half the page isn't empty as it is by default.
body {
  max-width: 1850px !important;
}
div.main-container {
  max-width: 1850px !important;
  width: 1850px !important;
  margin-left: auto !important;
  margin-right: auto !important;
}
.toc-content {
  padding-right: 0px;
  padding-left: 0px;
  margin-left: 300px;
  max-width: 1550px !important;
  width: 1550px !important;
}
```

```{r, include=FALSE}
library(BiocStyle)
knitr::opts_chunk$set(echo = TRUE, fig.align = "center")
```

# Introduction

Generating static reports from R markdown documents is a common way to distribute bioinformatics analyses with additional context and commentary. However, interactivity within these reports is limited. While one can embed minimally interactive figures in HTML reports, there is no way to alter the underlying data or plot by the end user.

[Shiny](https://shiny.rstudio.com/) provides an avenue to overcome this limitation, as [Shiny widgets](https://bookdown.org/yihui/rmarkdown/shiny-widgets.html) can be designed to interactively perform or display the results of an analysis. This comes with the downside that the Rmd file and necessary code must be hosted on a server that can run Shiny applications and interactive documents. [Options](https://www.rstudio.com/products/shiny/shiny-server/) include RStudio Connect, [shiny.apps.io](shiny.apps.io) (which has a free tier), and Shiny Server.

[iBET](https://github.com/j-andrews7/iBET) (interactive Bioconductor Exploratory Tools) is an R package that contains drop-in Shiny widgets that run and/or display the results of common analyses performed with Bioconductor R packages. These empower end-users to fully immerse themselves in the data through interactive alterations of plot and analysis parameters. Hopefully, it helps to alleviate some of the frustration that comes with frequent back-and-forth interactions to generate and alter figures.

# An Example

```{r, message = FALSE, warning = FALSE}
shh <- suppressPackageStartupMessages
shh(library("DESeq2"))
shh(library("PCAtools"))
shh(library("iBET"))
#shh(library("scRNAseq"))
#shh(library("scran"))
#shh(library("scuttle"))
#shh(library("scater"))

# Make a mock DESeq2Dataset.
dds <- makeExampleDESeqDataSet(n = 3000, m = 12, betaSD = 2)
dds$sample <- rownames(colData(dds))
dds$group <- rep(c("g1", "g2"), 6)
dds <- DESeq(dds)
vst <- assay(vst(dds))
```

## Interactive PCA via `PCAtools`

[PCAtools](https://bioconductor.org/packages/release/bioc/html/PCAtools.html) is a straight-forward package for principle component analysis. Note that filtering samples in the Metadata tab will re-run and re-plot the PCA.

```{r pca, message = FALSE, warning = FALSE}
shinyPCAtools(vst, metadata = colData(dds), annot.by = c("sample", "condition", "group"), color.by = "condition", shape.by = "group", scale = TRUE, height = 850)
```


### Larger (scRNA) Dataset

Works well on larger datasets, though it takes significant time for the PCA to run. Removing a much greater percentage of features is highly recommended.

```{r bigpca, message = FALSE, warning = FALSE, eval = FALSE}
# Not run
data <- ZeiselBrainData()
 
# # Remove genes expressed in few cells.
data <- data[rowMeans(counts(data) != 0) > 0.05, ]
data <- computeSumFactors(data, cluster = quickCluster(data))
data <- logNormCounts(data)
data$Cell.Type <- factor(data$level1class)

shinyPCAtools(logcounts(data), metadata = colData(data), annot.by = c("Cell.Type", "tissue", "age"), color.by = "Cell.Type", removeVar = 0.9, scale = TRUE)
```

## Interactive Differential Expression via `DESeq2`

This widget was heavily inspired by the `interactivate` function from the [InteractiveComplexHeatmap](https://bioconductor.org/packages/release/bioc/html/InteractiveComplexHeatmap.html) package. It wraps [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html) and will run differential expression analysis if not provided a results dataframe as well. 

Gene labels can be added to the MAplot and volcano plot by clicking a point. The labels can also be dragged around, though adding labels will reset the position, so it's recommended to add all labels prior to re-positioning them.

```{r de, message = FALSE, warning = FALSE}
shinyDESeq2(dds, annot.by = c("condition", "group"))
```

# SessionInfo

<details>

<summary>Click to expand</summary>

```{r, echo = FALSE}
sessionInfo()
```

</details>
