---
title: "iBET - interactive Bioinformatics Exploratory Tools"
author: 
- name: "Jared Andrews"
  email: jared.andrews@stjude.org
  affiliation: St. Jude Children's Research Hospital, Memphis, TN
date: "September 20th, 2021"
output: 
  html_document:
    code_folding: hide
    toc: true
    toc_depth: 5
    toc_float: true	 
runtime: shiny
vignette: >
  %\VignetteIndexEntry{iBET - Interactive Shiny Widgets for Bioinformatics Packages}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{css, echo=FALSE}
# This is to change width of the content so that half the page isn't empty as it is by default.
body {
  max-width: 1850px !important;
}
div.main-container {
  max-width: 1850px !important;
  width: 1850px !important;
  margin-left: auto !important;
  margin-right: auto !important;
}
.toc-content {
  padding-right: 0px;
  padding-left: 0px;
  margin-left: 300px;
  max-width: 1550px !important;
  width: 1550px !important;
}
```

```{r, include=FALSE}
library(BiocStyle)
knitr::opts_chunk$set(echo = TRUE, fig.align = "center", message = FALSE, warning = FALSE)
```

# Introduction

Generating static reports from R markdown documents is a common way to distribute bioinformatics analyses with additional context and commentary. However, interactivity within these reports is limited. While one can embed minimally interactive figures in HTML reports, there is no way to alter the underlying data or plot by the end user.

[Shiny](https://shiny.rstudio.com/) provides an avenue to overcome this limitation, as [Shiny widgets](https://bookdown.org/yihui/rmarkdown/shiny-widgets.html) can be designed to interactively perform or display the results of an analysis. This comes with the downside that the Rmd file and necessary code must be hosted on a server that can run Shiny applications and interactive documents. [Options](https://www.rstudio.com/products/shiny/shiny-server/) include RStudio Connect, [shiny.apps.io](shiny.apps.io) (which has a free tier), and Shiny Server.

[iBET](https://github.com/j-andrews7/iBET) (interactive Bioinformatics Exploratory Tools) is an R package that contains drop-in Shiny widgets that run and/or display the results of common analyses performed with bioinformatics R packages. These empower end-users to fully immerse themselves in the data through interactive alterations of plot and analysis parameters. Hopefully, it helps to alleviate some of the frustration that comes with frequent back-and-forth interactions to generate and alter figures. While these widgets are designed for drop-in use in Rmd notebooks, they also functions perfectly well as Shiny apps on their own when run from the console.

# The Widgets

First, let's load up some example data.

```{r, message = FALSE, warning = FALSE}
shh <- suppressPackageStartupMessages
shh(library("DESeq2"))
shh(library("PCAtools"))
shh(library("iBET"))
shh(library("edgeR"))
shh(library("airway"))
shh(library("msigdbr"))
shh(library("org.Hs.eg.db"))

data("airway")
airway$dex <- relevel(airway$dex, ref = "untrt")

# Get gene symbols.
ens <- rownames(airway)
symbols <- mapIds(org.Hs.eg.db, keys = ens,
  column = c('SYMBOL'), keytype = 'ENSEMBL')
symbols <- symbols[!is.na(symbols)]
symbols <- symbols[match(rownames(airway), names(symbols))]
rowData(airway)$SYMBOLS <- symbols
keep <- !is.na(rowData(airway)$SYMBOLS)
airway <- airway[keep,]

# Retrieve MSigDB Hallmark gene sets and convert to a named list.
hs.msig <- msigdbr(species = "Homo sapiens", category = "H")
hs.msig <- hs.msig %>% split(x = .$gene_symbol, f = .$gs_name)

dds <- DESeqDataSet(airway, design = ~ dex)
dds <- DESeq(dds)
vst <- assay(vst(dds))
rownames(vst) <- make.names(rowData(dds)$SYMBOLS, unique = TRUE)
```

## Interactive PCA via `PCAtools`

This interactive app provides provides several ways to assess variability within a dataset via [PCAtools](https://bioconductor.org/packages/release/bioc/html/PCAtools.html). It contains five tabs:

-   **biplot** - Contains an interactive biplot of PCA components (the traditional PCA plot). The PCA itself can be controlled via the PCA Settings box in the sidebar. The biplot can be adjusted in the Biplot Settings box.
-   **screeplot** - Contains a screeplot, which is used to assess the variability explained by each principal component. Useful for selecting the number of PCs to retain for downstream analyses. There are a few methods to select this number included in the Screeplot Settings box in the sidebar.
-   **eigencorplot** - This is a heatmap of variables by principal components to show how well each variable correlates with each PC. Currently, only numeric metadata variables may be used with it as factor variables depend on ordering and may yield confusing results.
-   **Distance Matrix** - This provides a distance matrix between all samples in the set. Annotations can easily be added to any or all axes.
-   **Metadata** - Contains a table of sample metadata. Can be used to filter the samples used in the PCA and shown in the biplot and distance matrices.

```{r pca, message = FALSE, warning = FALSE}
shinyPCAtools(vst, metadata = colData(dds), annot.by = c("cell", "dex"), 
              color.by = "dex", shape.by = "cell", scale = TRUE, height = 850)
```

### Larger (scRNA) Dataset

`shinyPCAtools` works well on larger datasets, though it takes significant time for the PCA to run. Removing a much greater percentage of features is highly recommended.

```{r bigpca, message = FALSE, warning = FALSE, eval = FALSE}
# Not run
shh(library("scRNAseq"))
shh(library("scran"))
shh(library("scuttle"))
shh(library("scater"))
data <- ZeiselBrainData()
 
# # Remove genes expressed in few cells.
data <- data[rowMeans(counts(data) != 0) > 0.05, ]
data <- computeSumFactors(data, cluster = quickCluster(data))
data <- logNormCounts(data)
data$Cell.Type <- factor(data$level1class)

shinyPCAtools(logcounts(data), metadata = colData(data), annot.by = c("Cell.Type", "tissue", "age"), 
              color.by = "Cell.Type", removeVar = 0.9, scale = TRUE)
```

## Interactive Differential Expression via `DESeq2`

This widget was heavily inspired by the `interactivate` function from the [InteractiveComplexHeatmap](https://bioconductor.org/packages/release/bioc/html/InteractiveComplexHeatmap.html) package. It wraps [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html) and will run differential expression analysis if not provided a results dataframe as well.

Gene labels can be added to the MAplot and volcano plot by clicking a point. The labels can also be dragged around, though adding labels will reset the position, so it's recommended to add all labels prior to re-positioning them. Gene sets can be highlighted easily if provided.

Multiple comparisons can also be provided for switching between analyses quickly.

```{r de, message = FALSE, warning = FALSE}
deseq.res1 <- results(dds, contrast = c("dex", "trt", "untrt"))
dds <- DESeqDataSet(airway, design = ~ cell)
dds <- DESeq(dds)
deseq.res2 <- results(dds, contrast = c("cell", "N080611", "N052611"))
deseq.res3 <- results(dds, contrast = c("cell", "N61311", "N080611"))
deseq.res4 <- results(dds, contrast = c("cell", "N080611", "N61311"))

deseq.res1$SYMBOLS <- make.names(rowData(dds)$SYMBOLS, unique = TRUE)
deseq.res2$SYMBOLS <- make.names(rowData(dds)$SYMBOLS, unique = TRUE)
deseq.res3$SYMBOLS <- make.names(rowData(dds)$SYMBOLS, unique = TRUE)
deseq.res4$SYMBOLS <- make.names(rowData(dds)$SYMBOLS, unique = TRUE)

rezzy <- list("trt v untrt" = as.data.frame(deseq.res1), 
            "N080611vN052611" = as.data.frame(deseq.res2), 
            "N61311vN080611" = as.data.frame(deseq.res3), 
            "N080611vN61311" = as.data.frame(deseq.res4))

shinyDESeq2(dds, res = rezzy, genesets = hs.msig, swap.rownames = "SYMBOLS", annot.by = c("cell", "dex"))
```

## Correlating DE Results

This widget creates scatter plots for arbitrary combinations of DE results. It takes a named list of dataframes as input and will plot scatter plots of log2 Fold Change values for each gene. It accepts up to 4 results dataframes and will generate plots for all combinations of them along with a regression line and correlation testing. Points are colored by significance in each DE analysis. Results dataframes from edgeR, limma, and DESeq2 will automatically have the appropriate columns used for plotting, but users can also provide the column names for their significance value or fold change column as necessary.

Gene labels can be added to a plot by clicking a point. The labels can also be dragged around, though adding labels to a plot will reset the label positions for said plot, so it's recommended to add all labels prior to re-positioning them.

```{r}
shinyDECorr(rezzy, gene.col = "SYMBOLS", genesets = hs.msig, height = 800)
```

# SessionInfo

<details>

<summary>

Click to expand

</summary>

```{r, echo = FALSE}
sessionInfo()
```

</details>
